presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddIntermediateAggregations.java

```diff
@@ -0,0 +1,191 @@
+ *   - RemoteExchange (GATHER)
+ *     - Aggregation (PARTIAL)
+ * </pre>
+ * To:
+ * <pre>
+ * - Aggregation (FINAL)
+ *   - LocalExchange (GATHER)
+ *     - Aggregation (INTERMEDIATE)
+ *       - LocalExchange (ARBITRARY)
+ *         - RemoteExchange (GATHER)
+ *           - Aggregation (INTERMEDIATE)
+ *             - LocalExchange (GATHER)
+ *               - Aggregation (PARTIAL)
+ * </pre>
+ * <p>
+ */
+public class AddIntermediateAggregations
+        implements Rule
+{
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!SystemSessionProperties.isEnableIntermediateAggregations(session)) {
+            return Optional.empty();
+        }
+
+        if (!(node instanceof AggregationNode)) {
+            return Optional.empty();
+        }
+
+        AggregationNode aggregation = (AggregationNode) node;
+
+        // Only consider FINAL un-grouped aggregations
+        if (aggregation.getStep() != AggregationNode.Step.FINAL || !aggregation.getGroupingKeys().isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<PlanNode> rewrittenSource = recurseThroughExchanges(lookup.resolve(aggregation.getSource()), lookup, idAllocator);
+
+        if (!rewrittenSource.isPresent()) {
+            return Optional.empty();
+        }
+
+        PlanNode source = rewrittenSource.get();
+
+        if (SystemSessionProperties.getTaskConcurrency(session) > 1) {
+            source = ExchangeNode.partitionedExchange(
+                    idAllocator.getNextId(),
+                    ExchangeNode.Scope.LOCAL,
+                    source,
+                    new PartitioningScheme(Partitioning.create(FIXED_ARBITRARY_DISTRIBUTION, ImmutableList.of()), source.getOutputSymbols()));
+            source = new AggregationNode(
+                    idAllocator.getNextId(),
+                    source,
+                    inputsAsOutputs(aggregation.getAssignments()),
+                    aggregation.getGroupingSets(),
+                    AggregationNode.Step.INTERMEDIATE,
+                    aggregation.getHashSymbol(),
+                    aggregation.getGroupIdSymbol());
+            source = ExchangeNode.gatheringExchange(idAllocator.getNextId(), ExchangeNode.Scope.LOCAL, source);
+        }
+
+        return Optional.of(ChildReplacer.replaceChildren(node, ImmutableList.of(source)));
+    }
+
+    /**
+     * Recurse through a series of preceding ExchangeNodes to find the preceding PARTIAL aggregation
+     */
+    private Optional<PlanNode> recurseThroughExchanges(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator)
+    {
+        if (node instanceof AggregationNode && ((AggregationNode) node).getStep() == AggregationNode.Step.PARTIAL) {
+            return Optional.of(addGatheringIntermediate((AggregationNode) node, idAllocator));
+        }
+
+        if (!(node instanceof ExchangeNode)) {
```

Comment:
FYI, pushing partial aggregation through UNION can lead to plans like:
```
AggFinal
  Union (RemoteExchange)
     ProjectNode
       AggPartial
         Source1 (...)
      AggPartial
        Source2 (...)
```

In that case (because of `ProjectNode`), this code will be unable to insert intermediate aggregations. I presume that this is not an issue for you. Maybe add a `TODO` for this?